### DISCUSSION TITLE
Consumed power and generated heat

### DISCUSSION THREAD

**Discussion by bmesuere** [raomin/ESPAltherma#174] [Q&A]
Is there a know way to get the "generated heat" values out of the Altherma unit (next to the consumed power, but this can easily be measured externally)? I would love to keep tabs on the COP of the unit without having to manually write down the values displayed on the unit every month.

I have checked the list of values on the wiki page but haven't found anything related.

**Comment by raomin** [comment]
Like this? https://github.com/raomin/ESPAltherma#calculating-cop
🙂

**Reply by bmesuere** [reply]
Not really, this is the real time COP (which is also valuable). I'm trying to monitor it over a longer period e.g., by week or month.

**Comment by alexdelli** [comment]
try to use Grafana

or use 

utility_meter:

  consumo_pompa_di_calore_giornaliero_kwh:
    source: sensor.shellyem_5e1757_channel_2_energy
    cycle: daily
  consumo_pompa_di_calore_mensile_kwh:
    source: sensor.shellyem_5e1757_channel_2_energy
    cycle: monthly

**Comment by 88gaspa88** [comment]
Hi,
To calculate the generated heat in KWH I created a couple of sensors in home assistant:
first one instantaneous heating power in W, with something like this:
      state: >
        {% set delta = states('sensor.temperatura_uscita_scambiatore_acqua_pdc') | float - states('sensor.temperatura_ingresso_scambiatore_acqua_pdc') | float%}
        {{ (delta * (states('sensor.portata_circuito_riscaldamento') | float) * 60 * 1.16) | round(2)}}

I have the top grade external unit so I can access the water temperature right before and after the water/gas heat exchanger, with this temperatures the power reading should be very accurate and stable, if you use the R1T and R4T values there will be some delays in the instantaneous readings, but the overall power should be accurate nevertheless.

once you have the instantaneous heat power sensor you can create a Riemann sum integral sensor in the helper page of HA, select trapezoidal rule, set as input sensor the power sensor just created and you are set.

If you then add the helper sensor to the energy dashboard you can see the daily, weekly etc heat generated right next to the power consumption (I'm using a shelly for power consumption)

If you want to do a step further you can use the utility meter integration to split the power consumption and heat generated between defrost phase, domestic hot water generation and heat generation.
![image](https://user-images.githubusercontent.com/110385116/201468678-cc62e59d-7ff4-461b-bf26-0fdd66c3cce1.png)

**Reply by alexdelli** [reply]
> Hi, To calculate the generated heat in KWH I created a couple of sensors in home assistant: first one instantaneous heating power in W, with something like this: state: > {% set delta = states('sensor.temperatura_uscita_scambiatore_acqua_pdc') | float - states('sensor.temperatura_ingresso_scambiatore_acqua_pdc') | float%} {{ (delta * (states('sensor.portata_circuito_riscaldamento') | float) * 60 * 1.16) | round(2)}}
> 
> I have the top grade external unit so I can access the water temperature right before and after the water/gas heat exchanger, with this temperatures the power reading should be very accurate and stable, if you use the R1T and R4T values there will be some delays in the instantaneous readings, but the overall power should be accurate nevertheless.
> 
> once you have the instantaneous heat power sensor you can create a Riemann sum integral sensor in the helper page of HA, select trapezoidal rule, set as input sensor the power sensor just created and you are set.
> 
> If you then add the helper sensor to the energy dashboard you can see the daily, weekly etc heat generated right next to the power consumption (I'm using a shelly for power consumption)
> 
> If you want to do a step further you can use the utility meter integration to split the power consumption and heat generated between defrost phase, domestic hot water generation and heat generation. ![image](https://user-images.githubusercontent.com/110385116/201468678-cc62e59d-7ff4-461b-bf26-0fdd66c3cce1.png)

Great job Gaspa
just a question: how to split the consumption/heat generated via utility meter? I cant undertand
thank you

**Reply by 88gaspa88** [reply]
> Great job Gaspa just a question: how to split the consumption/heat generated via utility meter? I cant undertand thank you
First you create an utility meter, then set as input the energy sensor (either the electric energy or the heat energy), set no cycle and under supported tariffs you can insert 4 different tariffs: "standby" "defrost" "heating" "DHW", the stand by is only for the electric energy, you are not going to have heat energy while the unit is in standby.

Once the utility sensors are created you have to create an automation to switch the active utility meter based on the unit status.
The triggers for the automation, and the conditions, will be the three way valve position, the unit status (fan only or heating) and the defrost status. Here is my automation for the electric energy, for heat energy will be the same but without stand by.
I think that something wrong may happen during defrost, because a negative energy will probably be generated, I presume that I'll find out probably in December.

`alias: Cambio modalita consumi PDC
description: ""
trigger:
  - platform: state
    entity_id:
      - sensor.defrost_pdc
  - platform: state
    entity_id:
      - sensor.richiesta_sanitario
  - platform: state
    entity_id:
      - sensor.modo_funzionamento
condition: []
action:
  - if:
      - condition: state
        entity_id: sensor.defrost_pdc
        state: "ON"
    then:
      - service: select.select_option
        data:
          option: scongelamento
        target:
          entity_id: select.pdc_energia_per_modalita

  - if:
      - condition: state
        entity_id: sensor.defrost_pdc
        state: "OFF"
      - condition: state
        entity_id: sensor.modo_funzionamento
        state: Heating
      - condition: state
        entity_id: sensor.richiesta_sanitario
        state: "ON"
    then:
      - service: select.select_option
        data:
          option: sanitario
        target:
          entity_id: select.pdc_energia_per_modalita

  - if:
      - condition: state
        entity_id: sensor.defrost_pdc
        state: "OFF"
      - condition: state
        entity_id: sensor.modo_funzionamento
        state: Heating
      - condition: state
        entity_id: sensor.richiesta_sanitario
        state: "OFF"
    then:
      - service: select.select_option
        data:
          option: riscaldamento
        target:
          entity_id: select.pdc_energia_per_modalita

  - if:
      - condition: state
        entity_id: sensor.modo_funzionamento
        state: Fan Only
    then:
      - service: select.select_option
        data:
          option: stand by
        target:
          entity_id: select.pdc_energia_per_modalita
mode: single`
